### DISCUSSION TITLE
DisallowConcurrentExecutionAttribute usage

### DISCUSSION THREAD

**Discussion by danies8** [quartznet/quartznet#2258] [Q&A] [answered]
Hi,
 DisallowConcurrentExecutionAttribute  - An attribute that marks a <see cref="IJob" /> class as one that must not have multiple
 instances executed concurrently (where instance is based-upon a <see cref="IJobDetail" />
 definition - or in other words based upon a <see cref="JobKey" />).

1. If I delete running JobKey and create same one by code when there is free, does this attrubute will block it?
My code:
`       // if route edit or scheduler edit, update DB with Up-to-date field to up-to-date
       // and reschedule the job by deleting job and schedule the job again.
       else if (routeData.ScheduleStatus == "Edit" && !MongoDbHelper.IsRouteRecordAlreadyUpdated(routeData.Id, "Up-to-date"))
       {
           try
           {
               var jobName = routeData.ScheduleId;
               var oldJobName = routeData.OldSchedulerId;
               var groupName = routeData.Id;
               var cronExpression = (route.Value.ScheData.CronExpression == null || route.Value.ScheData.CronExpression.Equals(string.Empty))
                            ? null : CronScheduleBuilder.CronSchedule(route.Value.ScheData.CronExpression);
               var startDate = route.Value.ScheData.ActiveStartDate;
               var endDate = route.Value.ScheData.ActiveEndDate;
               var startTime = route.Value.ScheData.ActiveStartTime;
               var endTime = route.Value.ScheData.ActiveEndTime;

               IJobDetail job = JobBuilder.Create<JobSchedulerHelper>()
                   .WithIdentity(jobName, groupName)
                    .Build();

               string sDate = $"{startDate} {startTime}";
               string eDate = $"{endDate} {endTime}";
               string formatter = null;
               if (startTime.Length > 5)
               {
                   formatter = "yyyy-MM-dd HH:mm:ss";
               }
               else
               {
                   formatter = "yyyy-MM-dd HH:mm";
               }
               var startDateTime = Convert.ToDateTime(DateTime.Parse(sDate).ToString(formatter));
               ITrigger newTrigger = null;
               if (cronExpression == null)// One time 
               {
                   _logger.Debug($"GUID:{emailGuid}, One time: startDateTime:{startDateTime}");
                   newTrigger = TriggerBuilder.Create().WithIdentity(jobName, groupName).StartAt(startDateTime.AddSeconds(30))
                               .Build();

               }
               else // Recuring
               {
                   if (endTime.Length > 5)
                   {
                       formatter = formatter = "yyyy-MM-dd HH:mm:ss";
                   }
                   else
                   {
                       formatter = "yyyy-MM-dd HH:mm";
                   }
                   var endDateTime = Convert.ToDateTime(DateTime.Parse(eDate).ToString(formatter));

                   newTrigger = TriggerBuilder.Create().WithIdentity(jobName, groupName).StartAt(startDateTime)
                  .WithCronSchedule(route.Value.ScheData.CronExpression, c => c.WithMisfireHandlingInstructionFireAndProceed()).EndAt(endDateTime).Build();

               }

               var jobKey = new JobKey(oldJobName, groupName);
               var executingJobs = await context.Scheduler.GetCurrentlyExecutingJobs();

               if (!executingJobs.Any(x => x.JobDetail.Key.Equals(jobKey)))
               {
                   //Update DB with Up-to-date field to Up-to-date;
                   MongoDbHelper.UpdateRouteRecord(routeData.Id,
                       DateTime.Now.ToString(ConstantsHelper.DateFormat), routeData.UserBy, "Up-to-date");
                   _logger.Debug($"GUID:{emailGuid}, route name: {routeData.Name} changes is up to date field from edit to up-to-date.");

                   var delete = await context.Scheduler.**DeleteJob**(new JobKey(oldJobName, groupName));
                   _logger.Debug($"GUID:{emailGuid}, route name: {routeData.Name} delete the  job ( old job name ={oldJobName},group name ={groupName}).");

                   var trigger = (IOperableTrigger)newTrigger;
                   trigger.StartTimeUtc = DateTimeOffset.UtcNow;
                   trigger.SetNextFireTimeUtc(null); //force recompute of nextFireTime based on trigger schedule. 
                   var offset = await context.Scheduler.**ScheduleJob**(job, newTrigger);
                   _logger.Debug($"GUID:{emailGuid}, route name: {routeData.Name} schedule the job(job name={jobName},group name ={groupName}).");
               }
               else
               {
                   _logger.Debug($"GUID:{emailGuid}, route name: {routeData.Name} is busy wait to next tick.");
                   continue;
               }
           }
           catch (Exception ex)
           {
               _logger.Error($"GUID:{emailGuid}, error in edit scheduler: shceduler ID = {route.Value.ScheduleId}, with name {route.Value.ScheduleName} , with route name {route.Value.Name}", ex);
               continue;
           }

       }
`

**Comment by da3dsoul** [comment]
What you are referring to seems more like rescheduling. DisallowConcurrentExecution is more for situations where multiple instances of the same type of job have been scheduled, likely with different job data.

For example, a DownloadLargeFileJob. It downloads a large file. Due to IO constraints or something, you want to queue 3 large files, but don't want them to run at the same time. They are different files, so the jobs details will be different. The JobKey might be the same (if you've set it via attribute or something), even though the jobs are different. DisallowConcurrentExecution makes the scheduler only allow one DownloadLargeFileJob to run at the same time (assuming they have the same JobKey).

**Comment by danies8** [comment, marked as answer]
@da3dsoul 
Thanks a lot.
